[16.0] [ADD] account_chart_update_multilang - new module

edlopen opened the pull request:
This new module extends the functionality of account_chart_update so that if the tax or fiscal position template has translated fields, it is passed to the final tax or fiscal position when applying the chart of accounts.

https://www.loom.com/share/ddc5e524a5f540e3ae680dc7f40d0090?sid=872b10d1-9a61-4d52-8452-91546a77ccbe

Although I am working to improve, please bear with me because of my English.

@Shide @yajo @rafaelbn please review, if you can.

@moduon MT-7597

Commit:
[16.0] account_chart_update - separate common data from actual tests

MT-7597

Commit:
[16.0][ADD] account_chart_update_multilang

MT-7597

edlopen commented:
I'll address the test error on Monday. I have no clue of why is failing at the moment.

yajo commented on account_chart_update_multilang/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "category": "Accounting",
+    "website": "https://github.com/OCA/account-financial-tools",
+    "author": "Moduon, Odoo Community Association (OCA)",
+    "maintainers": ["edlopen"],
```
```suggestion
    "maintainers": ["edlopen", "rafaelbn"],
```

yajo commented on account_chart_update_multilang/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+        "account_chart_update",
+        "l10n_multilang",
+    ],
+    "data": [],
```
```suggestion
```

yajo commented on account_chart_update_multilang/__manifest__.py:
```diff
@@ -0,0 +1,21 @@
+    "author": "Moduon, Odoo Community Association (OCA)",
+    "maintainers": ["edlopen"],
+    "license": "AGPL-3",
+    "application": False,
```
suggestion: make it auto-installable. If you happen to have all dependencies here, you probably want it.

```suggestion
    "auto_install": True,
```

(Before you ask, `application=False` is implicit)

yajo commented on account_chart_update_multilang/readme/DESCRIPTION.md:
```diff
@@ -0,0 +1 @@
+This module extends the functionality of account_chart_update so that if the tax or fiscal position template has translated fields, it is passed to the final tax or fiscal position when applying the chart of accounts.
```
```suggestion
This module extends the functionality of `account_chart_update` so that if taxes or fiscal position templates have translated fields, they are passed to the final taxes or fiscal positions when applying the chart of accounts.
```

yajo commented on account_chart_update_multilang/readme/USAGE.md:
```diff
@@ -0,0 +1,3 @@
+To use this module simply follow the steps in the USAGE section of the account_chart_update module.
```
```suggestion
To use this module simply follow the steps in the USAGE section of the `account_chart_update` module.
```

yajo commented on account_chart_update_multilang/readme/USAGE.md:
```diff
@@ -0,0 +1,3 @@
+To use this module simply follow the steps in the USAGE section of the account_chart_update module.
+
+You can find the documentation here: https://github.com/OCA/account-financial-tools/blob/16.0/account_chart_update/readme/USAGE.rst
```
Better point to the full readme:

```suggestion
You can find the documentation here: https://github.com/OCA/account-financial-tools/tree/16.0/account_chart_update#usage
```

yajo replied:
This is still missing.

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+        res = {}
+        langs = self.env["res.lang"].search([])
+        for key, _ in template._fields.items():
+            if not template._fields[key].translate:
```
```suggestion
        for key, field in template._fields.items():
            if not field.translate:
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+                continue
+            for lang in langs:
+                template_trans = getattr(template.with_context(lang=lang.code), key)
+                real_trans = getattr(real.with_context(lang=lang.code), key)
```
suggestion: when getting dynamic fields from a recordset, it's better to treat them as a dict, because that way the ORM will make sure you're getting a field and not another kind of attribute as it would be a function or other kind of object. Of course that doesn't affect here because you're already iterating over fields, but it's a good rule of thumb to remember and it will make code shorter:

```suggestion
                template_trans = template.with_context(lang=lang.code)[key]
                real_trans = real.with_context(lang=lang.code)[key]
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+
+    def diff_fields(self, template, real):
+        """Allow the comparison of template translations with real
+        translations."""
```
suggestion: it's good to follow [PEP 257](https://peps.python.org/pep-0257/) for docstrings:

```suggestion
        """Compare template and real translations too."""
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+        res = super()._update_taxes()
+        for wiz_tax in self.tax_ids.filtered(lambda x: x.type != "new"):
+            template, tax = wiz_tax.tax_id, wiz_tax.update_tax_id
+            for key, _ in self.diff_translate_fields(template, tax).items():
```
```suggestion
            for key in self.diff_translate_fields(template, tax):
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+    def _update_taxes(self):
+        """Update the taxes with their template translations."""
+        res = super()._update_taxes()
+        for wiz_tax in self.tax_ids.filtered(lambda x: x.type != "new"):
```
suggestion: don't double-iterate. In general terms, if you're gonna iterate over a recordset, filtering it just double-iterates it. In this particular case it won't affect performance very much. But yet again, it's a good rule of thumb:

```suggestion
        for wiz_tax in self.tax_ids:
            if wiz_tax.type != "new":
                continue
```

This way you iterate only once.

edlopen replied:
I see the point. But if i want to exclude "new" wiz_tax then it should be:

`if wiz_tax.type == "new"`

yajo replied:
Ah of course! Good catch

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+    def _update_fiscal_positions(self):
+        """Update the fiscal positions with their template translations."""
+        res = super()._update_fiscal_positions()
+        for wiz_fp in self.fiscal_position_ids.filtered(lambda x: x.type != "new"):
```
```suggestion
        for wiz_fp in self.fiscal_position_ids:
            if wiz_fp != "new":
                continue
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,54 @@
+        res = super()._update_fiscal_positions()
+        for wiz_fp in self.fiscal_position_ids.filtered(lambda x: x.type != "new"):
+            template, fp = wiz_fp.fiscal_position_id, wiz_fp.update_fiscal_position_id
+            for key, _ in self.diff_translate_fields(template, fp).items():
```
```suggestion
            for key in self.diff_translate_fields(template, fp):
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,58 @@
+                real_trans = getattr(real.with_context(lang=lang.code), key)
+                if template_trans != real_trans:
+                    res[key] = template[key]
+                    continue
```
It's pointless:

```suggestion
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,58 @@
+        """Update the fiscal positions with their template translations."""
+        res = super()._update_fiscal_positions()
+        for wiz_fp in self.fiscal_position_ids:
+            if wiz_fp == "new":
```
```suggestion
            if wiz_fp.type == "new":
```

yajo commented on account_chart_update/wizard/wizard_chart_update.py:
```diff
@@ -783,6 +783,22 @@ def fields_to_ignore(self, name):
         } | specials_mapping.get(name, set())
         return set(models.MAGIC_COLUMNS) | specials
 
+    @api.model
+    @tools.ormcache("name")
```
This is not an api.model function because it does need a record.

Besides, caching it won't save much time because the things it does are quite cheap.

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,57 @@
+        return self.env["res.lang"].search([("code", "!=", self.lang)]).mapped("code")
+
+    def _update_other_langs(self, templates):
+        for _, tpl_xmlid in templates.get_external_id().items():
```
```suggestion
        for tpl_xmlid in templates.get_external_id().values():
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,57 @@
+    def _diff_translate_fields(self, template, real):
+        """Find differences by comparing the translations of the fields."""
+        res = {}
+        to_include = self.fields_to_include(template._name)
+        for key in template._fields:
+            if not template._fields[key].translate or key not in to_include:
+                continue
```
suggestion: here you're iterating over all fields and then checking if they are in a list, which in turn loops behind the scenes to do that check.

It seems flipping the logic makes more sense:

```suggestion
        to_include = self.fields_to_include(template._name)
        for key in to_include:
            if not template._fields[key].translate:
                continue
```

yajo commented on account_chart_update_multilang/wizards/wizard_chart_update.py:
```diff
@@ -0,0 +1,57 @@
+            if not template._fields[key].translate or key not in to_include:
+                continue
+            for lang in self._other_langs():
+                template_trans = getattr(template.with_context(lang=lang), key)
+                real_trans = getattr(real.with_context(lang=lang), key)
+                if template_trans != real_trans:
+                    res[key] = template[key]
```
suggestion: Don't use getattr if you can avoid it, for security reasons. It was explained in some OXP security talk but I can't remember the year... In short, it makes sure the thing you're accessing is a field and not a method, property or something else.

```suggestion
                template_trans = template.with_context(lang=lang)[key]
                real_trans = real.with_context(lang=lang)[key]
                if template_trans != real_trans:
                    res[key] = template[key]
```

yajo commented:
I've diagnosed a bit more the problem of not being able to translate notes in fiscal positions. It's due to how the html translation algorithm works. Pushing a wip fix. Could you set this as draft please?

yajo committed:
[FIX] account_chart_update_multilang: proper translating of HTML fields

The algorithm to translate them is done by terms, which is a bit clunky for our use case.

However, since `account.fiscal.position.template`'s note field is Text, not Html, and that's the one we care about, we can predict that the Html version will just wrap it with a <p> tag.

Well, in short, it's ugly but works.

MT-7597

yajo commented:
Last push should fix the problem.

yajo reviewed (approved):
Reviewed in real DB and it works.

rafaelbn reviewed (approved):
/ocabot merge patch

👍🏼 Tested

Commit:
[FIX] account_chart_update_multilang: hide default language in wizard and ditch decorators as mentioned in code review

edlopen commented:
@rafaelbn it seems this is ready to merge now! All issues and errors have been solved.

rafaelbn commented:
/ocabot merge patch
